src/hdmf/common/resources.py

```diff
@@ -165,6 +166,57 @@ def __init__(self, **kwargs):
         self.object_keys = kwargs['object_keys'] or ObjectKeyTable()
         self.type_map = kwargs['type_map'] or get_type_map()
 
+    @staticmethod
+    def assert_external_resources_equal(left, right, check_dtype=True):
+        """
+        Compare that the keys, resources, entities, objects, and object_keys tables match
+
+        :param left: ExternalResources object to compare with right
+        :param right: ExternalResources object to compare with left
+        :param check_dtype: Enforce strict checking of dtypes. Dtypes may be different
+            for example for ids, where depending on how the data was saved
+            ids may change from int64 to int32. (Default: True)
+        :returns: The function returns True if all values match. If mismatches are found,
+            AssertionError will be raised.
+        :raises AssertionError: Raised if any differences are found. The function collects
+            all differences into a single error so that the assertion will indicate
+            all found differences.
+        """
+        errors = []
+        try:
+            pd.testing.assert_frame_equal(left.keys.to_dataframe(),
+                                          right.keys.to_dataframe(),
+                                          check_dtype=check_dtype)
+        except AssertionError as e:
+            errors.append(e)
+        try:
+            pd.testing.assert_frame_equal(left.objects.to_dataframe(),
+                                          right.objects.to_dataframe(),
+                                          check_dtype=check_dtype)
+        except AssertionError as e:
+            errors.append(e)
+        try:
+            pd.testing.assert_frame_equal(left.resources.to_dataframe(),
+                                          right.resources.to_dataframe(),
+                                          check_dtype=check_dtype)
+        except AssertionError as e:
+            errors.append(e)
+        try:
+            pd.testing.assert_frame_equal(left.entities.to_dataframe(),
+                                          right.entities.to_dataframe(),
+                                          check_dtype=check_dtype)
+        except AssertionError as e:
+            errors.append(e)
+        try:
+            pd.testing.assert_frame_equal(left.object_keys.to_dataframe(),
+                                          right.object_keys.to_dataframe(),
+                                          check_dtype=check_dtype)
+        finally:
+            if len(errors) > 0:
+                msg = ''.join(str(e)+"\n\n" for e in errors)
+                raise AssertionError(msg)
```

Comment:
Could you please explain this change? Why remove the `except` statement? Does this `finally` run only on the last `try` block?

Reply:
@mavaylon1 I'm still confused about this block of code and the change. Could you please explain it? It might be worth leaving a comment for future us there

Reply:
@rly The line removed was an except that was redundant. The prior except/try blocks accounted for all the tables. The finally just says "run this if statement always" and it will run after all the prior try blocks.

Reply:
The part that is confusing to me is that this try block does not have an except at all, i.e., the error that may be raised by this assert is not being recorded. Also, it is not clear to me why this code is part of the finally statement, rather than just code after the except. I.e., I would have expected something like:

```
try:
    pd.testing.assert_frame_equal(left.object_keys.to_dataframe(),
                                    right.object_keys.to_dataframe(),
                                    check_dtype=check_dtype)
except AssertionError as e:
    errors.append(e)
   
if len(errors) > 0:
    msg = ''.join(str(e)+"\n\n" for e in errors)
    raise AssertionError(msg)
```

Having the code in the finally statement is in principle fine, but I don't think it is necessary. However, without the except part, the error message from the last assert won't be recorded in the AssertionError that is being raised by the function

Reply:
Ahh I see what you're saying. I'll fix it now.